### ISSUE TITLE
Cross signing fails key match after several times of key mismatch

### ISSUE THREAD

**Issue by bitfriend** [matrix-org/matrix-rust-sdk#3216]
I was testing verification.
Once any device is verified, we can try verification for that device no more.
Single solution is to logout that device and clear verification history.
So I responded key mismatch to avoid logout in every time.
```rs
sas.mismatch().await;
```
I could finish handling of all events prior to `SasState::Confirmed` via several times of mismatch.
I tried key match, but it always failed with [CancelCode::KeyMismatch](https://github.com/matrix-org/matrix-rust-sdk/blob/1f52aca210e237656f5ea56544f4279a28298671/crates/matrix-sdk-crypto/src/verification/mod.rs#L512).
```rs
sas.confirm().await;
```

**Comment by bitfriend** [comment]
I got some idea from [this comment](https://github.com/matrix-org/matrix-rust-sdk/blob/1f52aca210e237656f5ea56544f4279a28298671/crates/matrix-sdk-crypto/src/verification/mod.rs#L510-L511).

I did re-login that device and regenerate session & device id.
This fixed my issue but I don't know how it could fix this issue.

I wonder if I should recommend re-login solution when user complaints this issue.

**Comment by bnjbvr** [comment]
Thanks for opening an issue, but still not sure what's going on here; could you write a failing test, please?

**Comment by bitfriend** [comment]
I wrote the procedure [here](https://github.com/matrix-org/matrix-rust-sdk/issues/3216#issue-2186876902).
Do you want test log?

**Comment by bnjbvr** [comment]
No, what I'd really like is a Rust `#[test]` or `#[async_test]` that shows all the steps required to cause the bug, and what you'd expect to see (using assertions).
